mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,716 @@
+            phantom: Default::default(),
+        }
+    }
+}
+
+struct FetchingTransactionsData<Coin, Storage> {
+    /// The list of addresses for those we have requested [`UpdatingUnconfirmedTxes::all_tx_ids_with_height`] TX hashses
+    /// at the `FetchingTxHashes` state.
+    address: String,
+    /// Denom - Ticker
+    active_assets: HashMap<String, String>,
+    phantom: std::marker::PhantomData<(Coin, Storage)>,
+}
+
+impl<Coin, Storage> FetchingTransactionsData<Coin, Storage> {
+    fn new(address: String, active_assets: HashMap<String, String>) -> Self {
+        FetchingTransactionsData {
+            address,
+            active_assets,
+            phantom: Default::default(),
+        }
+    }
+}
+
+impl<Coin, Storage> TransitionFrom<TendermintInit<Coin, Storage>> for Stopped<Coin, Storage> {}
+impl<Coin, Storage> TransitionFrom<TendermintInit<Coin, Storage>> for FetchingTransactionsData<Coin, Storage> {}
+impl<Coin, Storage> TransitionFrom<WaitForHistoryUpdateTrigger<Coin, Storage>> for Stopped<Coin, Storage> {}
+impl<Coin, Storage> TransitionFrom<FetchingTransactionsData<Coin, Storage>> for Stopped<Coin, Storage> {}
+
+impl<Coin, Storage> TransitionFrom<WaitForHistoryUpdateTrigger<Coin, Storage>>
+    for FetchingTransactionsData<Coin, Storage>
+{
+}
+
+impl<Coin, Storage> TransitionFrom<FetchingTransactionsData<Coin, Storage>>
+    for WaitForHistoryUpdateTrigger<Coin, Storage>
+{
+}
+
+#[async_trait]
+impl<Coin, Storage> State for WaitForHistoryUpdateTrigger<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        loop {
+            Timer::sleep(30.).await;
+
+            let ctx_balances = ctx.balances.clone();
+
+            let balances = match ctx.coin.all_balances().await {
+                Ok(balances) => balances,
+                Err(e) => return Self::change_state(Stopped::unknown(e.to_string())),
+            };
+
+            if balances != ctx_balances {
+                if let Err(e) = ensure_storage_init(&ctx.coin.all_wallet_ids(), &ctx.storage).await {
+                    return Self::change_state(Stopped::storage_error(e));
+                }
+                // Update balances
+                ctx.balances = balances;
+
+                return Self::change_state(FetchingTransactionsData::new(
+                    ctx.coin.my_address().expect("my_address can't fail"),
+                    ctx.coin.get_denom_and_ticker_map(),
+                ));
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl<Coin, Storage> State for FetchingTransactionsData<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    // TODO
+    // - claim/create htlcs will display twice
```

Comment:
Is this TODO still actual?

Reply:
Ah, yes. I forgot to ask this. So create/claim transactions are 2 different transactions on the chain. For example, when we do single swap with IRIS, there will be 2 transactions(create and claim) we will get from tx_search. I am not sure if we should insert both or one(if one, which) of them into history.

Reply:
> I understood that create_htlc txes are duplicated their selves
No, they are proceed only once.

What I mean is after swap operation, if you send 1 `IRIS`, it will show 1 `IRIS` sent twice one as create htlc and one as claim htlc. Which might cause confusion for user on gui side. Am I wrong?

Reply:
Alright then I can remove this TODO comment since it's working as you expect it.